### ISSUE TITLE
Host a TensorBuilder patch for TensorFlow

### ISSUE THREAD

**Issue by cgarciae** [tensorflow/tensorflow#2750]
Hi,

>[TensorBuilder](https://github.com/cgarciae/tensorbuilder) is light-weight extensible library that enables you to easily create complex deep neural networks using functions from any Tensor-based library through a functional fluent immutable API based on the Builder Pattern. As a side effect, TensorBuilder is a library that gives expressive power to any Tensor-based library that decide to implement a TensorBuilder patch.

TensorBuilder hosts a patch for `tensorflow`, using this patch you can rewrite the code like this

```
import tensorflow as tf

x = tf.placeholder(tf.float32, shape=[None, 5])
keep_prob = tf.placeholder(tf.float32)

net = tf.contrib.layers.fully_connected(x, 10, activation_fn=tf.nn.tanh) # tanh(x * w + b)
net = tf.nn.dropout(net, keep_prob) # dropout(x, keep_prob)
h =  tf.contrib.layers.fully_connected(net, 3, activation_fn=tf.nn.softmax) # softmax(x * w + b)
```
as this 
```
import tensorflow as tf
import tensorbuilder as tb
import tensorbuilder.patches.tensorflow.slim

x = tf.placeholder(tf.float32, shape=[None, 5])
keep_prob = tf.placeholder(tf.float32)

h = (
    x.builder()
    .fully_connected(10, activation_fn=tf.nn.tanh)
    .map(tf.nn.dropout, keep_prob)
    .fully_connected(3, activation_fn=tf.nn.softmax)
    .tensor()
)
```
or even like this using a stronger patch
```
import tensorflow as tf
import tensorbuilder as tb
import tensorbuilder.patches.tensorflow.patch

x = tf.placeholder(tf.float32, shape=[None, 5])
keep_prob = tf.placeholder(tf.float32)

h = (
    x.builder()
    .tanh_layer(10) 
    .dropout(keep_prob)
    .softmax_layer(3)
    .tensor()
)
```
The patch for the `tensorflow` library currently has full support for branching, see this example with the strong patch
```python
import tensorflow as tf
import tensorbuilder as tb
import tensorbuilder.patches.tensorflow.patch

x = tf.placeholder(tf.float32, shape=[None, 5])
keep_prob = tf.placeholder(tf.float32)

h = (
    x.builder()
    .fully_connected(10)
    .branch(lambda root:
    [
        root
        .relu_layer(3)
    ,
        root
        .tanh_layer(9)
        .branch(lambda root2:
        [
          root2
          .sigmoid_layer(6)
        ,
          root2
          .dropout(keep_prob)
          .softmax_layer(8)
        ])
    ])
    .sigmoid_layer(6) #<== This connects all the branches to a single output layer
    .tensor()
)
```
TensorBuilder even has a DSL that you get for free, so you could rewrite the previous as
```
import tensorflow as tf
import tensorbuilder as tb
import tensorbuilder.patches.tensorflow.patch
import tensorbuilder.dsl as dl #<== Notice the alias

x = tf.placeholder(tf.float32, shape=[None, 5])
keep_prob = tf.placeholder(tf.float32)

h = x.builder().pipe(
    dl.fully_connected(10),
    [
        dl.relu_layer(3)
    ,
        (dl.tanh_layer(9),
        [
            dl.sigmoid_layer(6)
        ,
            dl
            .dropout(keep_prob)
            .softmax_layer(8)
        ])
    ],
    dl.sigmoid_layer(6)
    .tensor()
)
```
### Invitation
I'd like to invite you to eventually host the TensorFlow patch for TensorBuilder on the TensorFlow library, because TensorBuilder really is just an empty shell, it doesn't even know about tensors, its a sort of functional tool for making non-composable/fluent APIs into composable/fluent APIs (in e.g. Elixir or Elm you wouldn't even need it), but it benefits a lot from libraries, and I am sure that tensorflow users will benefit a lot from it too.

**Comment by girving** [comment]
@cgarciae That sounds like a great contribution.  What do you mean by hosting a patch?  It seems like it would be best as a separate repository.

**Comment by cgarciae** [comment]
@girving The proposal is to eventually move the module `tensorbuilder.patches.tensorflow` out of TensorBuilder to a repo managed by the TensorFlow team, that way TensorBuilder can concentrate in being a wrapper that gives nice syntax to Tensor-based libraries.

**Comment by cgarciae** [comment]
Here is the code that patches all the functions in the `tf.nn` as methods of the `Builder` class, just so you have a taste of the kind of code involved
```
import tensorflow as tf
import tensorbuilder as tb
import inspect

for _f_name, f in inspect.getmembers(tf.nn, inspect.isfunction):
    tb.Builder.register_map_method(f, "tensorflow.nn")
```

**Comment by girving** [comment]
@cgarciae We wouldn't be able to support it, but it's possible it could go in `tf.contrib`.  @martinwicke What do you think?

**Comment by martinwicke** [comment]
Since we cannot support it (and we would like to avoid confusion about that), it should not go into contrib. For now, keeping it in your repo makes the most sense. This is somewhat similar to PrettyTensor, but the DSL aspect is certainly nice -- it's been requested a lot.

**Comment by cgarciae** [comment]
1. I'll maintain it for the moment but leave the invitation open.
2. @martinwicke Yeah, its actually inspired by PrettyTensor (initially) but with some changes, the original README had this comment

>TensorBuilder takes inspiration from [prettytensor](https://github.com/google/prettytensor) but its internals are simpler, its API is smaller but equally powerfull, its branching mechanism is more expresive and doesn't break the fluent API, and its immutable nature helps avoid a lot of conceptual complexity.

A lot of problems I found with PrettyTensor is that its API was mutable, where as Tensor operations are immutable. So TensorBuilder is immutable, this makes creating an arbitrary number of branches trivial, you also don't need the Templates concept to specify a procedure, you just use a plain old python function from Builder to Builder.

In the last release I removed the core functionality for creating layers and delegated that to libraries through patches. Now TensorBuilder's focus is to enable Tensor-based libraries with nice syntax, branching capabilities, and the DSL.
